### DISCUSSION TITLE
How to get file path to transformed image?

### DISCUSSION THREAD

**Discussion by wsydney76** [craftcms/cms#11408] [Ideas] [answered]
Use case: Create and copy special transforms of selected images and package them into a native app, using a local transform filesystem.

This gets more complicated in Craft 4, eg. in Craft 3 there was `Craft::$app->assetTransforms->getTransformSubpath()`, that is now a protected method in the `ImageTransformer` class.

Thrown together some wild things in a behavior, learned a lot I never wanted to know, probably missing better approaches, finally  feeling uncomfortable with doing all this in detail by myself.

```
/**
     * Workaround because some file system methods are protected in Craft 4
     *
     * Format: <Transform file system path>/<opt. subpath>/<opt. folder(s)>/<transform string>/<opt. image.id>/<filename>
     *
     * @param $transform
     * @return string|null
     * @throws ImageTransformException
     * @throws InvalidConfigException
     */
    public function getImageTransformPath(string|array $transform): ?string
    {
        /** @var Asset $image */
        $image = $this->owner;

        // Get base path of transform filesystem
        $path = App::parseEnv($image->volume->getTransformFs()->path) . DIRECTORY_SEPARATOR;

        // Add subpath, if defined in volume
        if ($image->volume->transformSubpath) {
            $path .= $image->volume->transformSubpath . DIRECTORY_SEPARATOR;
        }

        // If image is in a subfolder, add path to the folder (has trailing slash ??)
        $folder = $image->folder->path;
        if ($folder) {
            $path .= $folder;
        }

        // Get transform index record
        /** @var ImageTransformer $transformer */
        $transformer = Craft::$app->imageTransforms->getImageTransformer(ImageTransformer::class);
        $transformIndex = $transformer->getTransformIndex($image, $transform);

        // Add transform string to path (_transformName or _<width>x<height>_<mode>_<pos>_none)
        $path .= $transformIndex->transformString . DIRECTORY_SEPARATOR;

        // Asset is in subfolder <id> if it has a different filename than original image (if format is different)
        if ($image->filename != $transformIndex->filename) {
            $path .= $image->id . DIRECTORY_SEPARATOR;
        }

        // Finally: Add filename of transformed image
        $path .= $transformIndex->filename;

        if (!file_exists($path)) {
            return null;
        }

        return $path;
    }
```

So it would be great to have a public method like `::getTransformedImagePath($asset, $transform)`

/Mel

PS.
Finally cheated by just replacing `@web` with `@webroot`, but that won't work with other setups e.g. with server aliases.

And it doesn't feel clean to use http calls.

**Comment by brandonkelly** [comment, marked as answer]
We should make this easier down the road for sure. For now, you can expose `getTransformSubpath()` as a public method by providing your own `ImageTransformer` class that extends `craft\imagetransforms\ImageTransformer`, like so:

```php
namespace my\namespace;

use craft\elements\Asset;
use craft\imagetransforms\ImageTransformer;
use craft\models\ImageTransformIndex;

class MyImageTransformer extends ImageTransformer
{
    public function getTransformSubpath(Asset $asset, ImageTransformIndex $transformIndex): string
    {
        return parent::getTransformSubpath($asset, $transformIndex);
    }
}
```

Then you can have Craft use your own `ImageTransformer` class instead of `carft\imagetransforms\ImageTransformer` using [dependency injection](https://www.yiiframework.com/doc/guide/2.0/en/concept-di-container) from a module’s `init()` method:

```php
namespace my\namespace;

use Craft;
use craft\imagetransforms\ImageTransformer;
use craft\models\ImageTransformIndex;

// ...

public function init()
{
    // ...

    Craft::$container->set(
        ImageTransformer::class,
        fn($container, $params, $config) => new MyImageTransformer($config),
    );
}
```

**Reply by wsydney76** [reply]
@brandonkelly Thanks, that works perfect.

Just out of curiosity (no priority): Does this approach (DI) work for every Craft class or are there limitations?

At first glance it looks like an alternative to using behaviors, adding the possibility to override methods?

**Reply by brandonkelly** [reply]
It only works in cases where the class is always instantiated via `Craft::createObject()`. I had to verify that was the case with `ImageTransformer` before offering it as a suggestion.

If you do come across something you’d like to override with the DI container, which is being instantiated directly (`new SomeClass()`), feel free to request that we switch over to `Craft::createObject()`. It’s a non-breaking change to do.
